Extend Label SDK: Support numbers and bools and add label to errors

gregkalapos opened the pull request:
Solves #967, #788, #473, #191

Prior to this we only had
- `ITransaction.Labels` as `Dictionary<string,string>`
- `ISpan.Labels` as `Dictionary<string,string>`

This PR adds new methods to add labels to both `ISpan` and `ITransaction` (#967, #788):

```
public void SetLabel(string key, string value);
public void SetLabel(string key, bool value);
public void SetLabel(string key, double value);
public void SetLabel(string key, int value);
public void SetLabel(string key, long value);
public void SetLabel(string key, decimal value);
```

and keeps the dictionary with a backwards compatible change (the dictionary can be deprecated long term):
- `ITransaction.Labels` as ~`Dictionary<string,object>`~ `Dictionary<string,object>` (no change)
- `ISpan.Labels` as ~`Dictionary<string,object>`~ `Dictionary<string,object>` (no change)

Additionally it also adds an optional parameter to pass labels to errors both on `ITransaction` and `ISpan` #473:

```
ITransaction.CaptureError("Test Error", "TestMethod", new StackTrace().GetFrames(),
	labels: new Dictionary<string, Label> { { "stringLabel", "test" } });
```

The `labels` parameter will be added to the labels on the error, which contains all the labels on the given transaction plus the ones passed to the `CaptureError` and `CaptureException` methods

Prior to the PR every `Error.Context` was just a reference to the existing `Transaction.Context`. The problem with this was that changes on one of them would be reflected on the other - these 2 `Context` instances should be decoupled (according to spec - and that's also what other agents do). This PR implements this with adding DeepCopy to `Context` - each time an `Error` is created, it deep-copies `Transaction.Context`  #191.

gregkalapos commented on src/Elastic.Apm/Api/IExecutionSegment.cs:
```diff
@@ -40,7 +40,7 @@ public interface IExecutionSegment
 		/// </summary>
 		/// <exception cref="ArgumentException"><c>null</c> as key is not allowed.</exception>
-		Dictionary<string, string> Labels { get; }
+		Dictionary<string, object> Labels { get; }
```
This isn't very nice, but I want to keep backwards compatibility - this way things like (current API usage) this will still compile and work:

```
ISpan.Labels["foo"] = "bar"
``` 

Long term I feel it'd be better to just depreciate the `Dictionary` from the API.

russcam replied:
Could a `Tag` (from what I see, tag is what labels really are?) type be introduced, that constrains the inputs only to 

- string
- bool
- double
- int
- long
- decimal

For example,

```c#
[JsonConverter(typeof(TagConverter))]
public class Tag
{
	internal object _object;
	
	public Tag(string value) => _object = value;
	
	public Tag(bool value) => _object = value;
	
	public Tag(double value) => _object = value;
	
	public Tag(int value) => _object = value;
	
	public Tag(long value) => _object = value;
	
	public Tag(decimal value) => _object = value;
	
	public static implicit operator Tag(string value) => new Tag(value);
	
	public static implicit operator Tag(bool value) => new Tag(value);
	
	public static implicit operator Tag(double value) => new Tag(value);
	
	public static implicit operator Tag(int value) => new Tag(value);
	
	public static implicit operator Tag(long value) => new Tag(value);
	
	public static implicit operator Tag(decimal value) => new Tag(value);
}

public class TagConverter : JsonConverter
{
	public override bool CanConvert(Type objectType) => typeof(Tag).IsAssignableFrom(objectType);

	public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
	{
		switch(reader.TokenType)
		{ 
			case JsonToken.Boolean:
				return new Tag((bool)reader.Value);
			case JsonToken.Float:
				return new Tag((double)reader.Value);
			case JsonToken.Integer:
				return new Tag((long)reader.Value);
			case JsonToken.String:
				return new Tag((string)reader.Value);
			default:
				throw new JsonException($"cannot deserialize {nameof(Tag)} from token type {reader.TokenType}");
		}
	}

	public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) =>
		writer.WriteValue(((Tag)value)._object);
}
```

This would maintain API compatibility and also constrain tags/labels to the types supported by APM server

gregkalapos replied:
I really liked this idea and I also implemented it (as discussed on zoom I named the class `Label`, but I use the same idea -  and I pushed it to this PR.

I ran into one (I think) fairly big problem: This indeed makes it a nice API and it's backward-compatible with things like `ISpan.Labels["foo"] = "bar"`, but once you want to read the value it becomes a breaking change:

Prior to this, here is how you read the value:
```
var myValue = `ISpan.Labels["foo"] ; // here `myValue` is a string
```

But once we add the `Label` type, `myValue` won't be a `string` anymore, it'll be a `Label` and in order to read the value you need to write:

```
var myValue = `ISpan.Labels["foo"].Value
```

This is definitely not the standard use-case (I expect users just to write into the `Labels` and never really read it in the code later), nevertheless this was totally possible with the old API and we'd break that.  ...and this change also breaks our tests from the same reason, so we also need to assert with `.Value` once we introduce the new `Label` type.

So, right now the suggestion is contained in the PR, but I have slight preference to not add this class because of this problem. Sad to not have the cool API that'd promote the possible types, but I'd rather not break the reading use case.  Any idea or thoughts on this?

russcam replied:
One could look at it from the perspective that the agent's API does not conform to the APM server spec for labels, because it only supports strings, and thus this is a bug that should be addressed. It is unfortunate that the fix is a breaking change.

I think a working around of this breaking change is probably not worth the effort and adds more complexity to the API. For example, an implicit operator of `Label` -> `string` could be added

```c#
public class Label
{
    public static implicit operator string(Label value) => value.Value.ToString();
}
```

`myValue` would still be of type `Label`, but would convert to `string` in cases where a `string` value is expected e.g.

```c#
void Main()
{
	var foo = new Foo();
	
	foo.Labels.Add("label_1", "one");
	foo.Labels.Add("label_2", 2);
	
        // conversion to string on assignment
	string string1 = foo.Labels["label_1"];
	Write(string1);
	
        // conversion to string on pass to method
	var label1 = foo.Labels["label_1"];
	Write(label1);
}

public void Write(string value) => Console.WriteLine(value);

public class Foo
{
	private Dictionary<string, Label> _labels = new Dictionary<string, Label>();	
	public Dictionary<string, Label> Labels => _labels;
}
```

What makes me uncomfortable with this approach is that any value `Label` holds would be converted to string, rather than handled as the type it is.

I personally think it would be better to introduce the breaking change, and bump the major version, if the value of supporting multiple types for label values is important; I suspect that this might be useful to have for performing certain types of queries such as range queries on numeric values.

gregkalapos replied:
We had a chat about this over zoom. We concluded that actually all the suggestions are breaking changes (no matter if it's  `Dictionary<string,object>` or `Dictionary<string,Label>`, so we decided that I'll do another round on this and try to keep the API as `Dictionary<string,string>` - that's to only way to be sure we don't break user code.

I implemented this and pushed to this PR.

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-dotnet_982.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #982 opened]`

* Start Time: 2020-10-19T21:31:49.723+0000

* Duration: 60 min 57 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 128  |
| Passed       | 14115  |
| Skipped      | 0 |
| Total        | 14243   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![128](https://img.shields.io/badge/128%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//tests)
  <details><summary>Expand to view the tests failures</summary>
  <p>
    > Show only the first 10 test failures
    
* **Name**: `Initializing / Parallel / Linux / Test / AspNetCoreDiagnosticListenerTest.Elastic.Apm.AspNetCore.Tests.AspNetCoreDiagnosticListenerTest.TestJsonBodyRetrievalOnRequestFailureInAspNetCore(useOnlyDiagnosticSource: True) – Elastic.Apm.AspNetCore.Tests.AspNetCoreDiagnosticListenerTest`
  * Age: 1
  * Duration: 0.252
  * Error Details: Expected context.Request.Body to be "{"id" : "1"}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / AspNetCoreDiagnosticListenerTest.Elastic.Apm.AspNetCore.Tests.AspNetCoreDiagnosticListenerTest.TestJsonBodyRetrievalOnRequestFailureInAspNetCore(useOnlyDiagnosticSource: False) – Elastic.Apm.AspNetCore.Tests.AspNetCoreDiagnosticListenerTest`
  * Age: 1
  * Duration: 0.102
  * Error Details: Expected context.Request.Body to be "{"id" : "1"}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.338
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 0, startCfgVariant: OptionsTestVariant{CaptureBody: off, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, IsSampled: False}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.217
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 1, startCfgVariant: OptionsTestVariant{CaptureBody: off, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, IsSampled: True}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.214
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 2, startCfgVariant: OptionsTestVariant{CaptureBody: off, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: False}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.221
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 3, startCfgVariant: OptionsTestVariant{CaptureBody: off, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.222
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 4, startCfgVariant: OptionsTestVariant{CaptureBody: all, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, IsSampled: False}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.283
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 5, startCfgVariant: OptionsTestVariant{CaptureBody: all, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, IsSampled: True}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.238
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 6, startCfgVariant: OptionsTestVariant{CaptureBody: all, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: False}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
* **Name**: `Initializing / Parallel / Linux / Test / BodyCapturingTests.Elastic.Apm.AspNetCore.Tests.BodyCapturingTests.Options_changed_after_start – Elastic.Apm.AspNetCore.Tests.BodyCapturingTests`
  * Age: 1
  * Duration: 0.218
  * Error Details: Expected error.Context.Request.Body to be "{startCfgVariantIndex: 7, startCfgVariant: OptionsTestVariant{CaptureBody: all, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, updateCfgVariantIndex: 11, updateCfgVariant: OptionsTestVariant{CaptureBody: errors, CaptureBodyContentTypes: application/x-www-form-urlencoded*, text/*, application/json*, application/xml*, application/my-custom-content, IsSampled: True}, isError: True}", but found "[REDACTED]".
    
  </p>
  </details>


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-982/detail/PR-982/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  
  * **Name**: `Test IIS`
    * Description: .ci/windows/test-iis.bat
    
    * Duration: 13 min 13 sec
    
    * Start Time: 2020-10-19T22:18:23.941+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-982/runs/1/steps/393/log)
    
  
  * **Name**: `Build`
    * Description: .ci/windows/dotnet.bat
    
    * Duration: 0 min 23 sec
    
    * Start Time: 2020-10-19T21:42:49.773+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-982/runs/1/steps/216/log)
    
  
  * **Name**: `Test & coverage`
    * Description: .ci/windows/test.bat
    
    * Duration: 11 min 27 sec
    
    * Start Time: 2020-10-19T21:49:47.213+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-982/runs/1/steps/249/log)
    
  
  * **Name**: `General Build Step`
    * Description: [2020-10-19T22:01:18.121Z] Archiving artifacts
script returned exit code 1
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-10-19T22:01:18.106+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-982/runs/1/steps/327/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-10-19T22:31:34.201Z]       Route matched with {action = "Post", controller = "IntakeV2Events"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] Post() on controller Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController (Elastic.Apm.Tests.MockApmServer).
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
[2020-10-19T22:31:34.201Z]       Executing action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) - Validation state: Valid
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.201Z]       Executed action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer), returned result Microsoft.AspNetCore.Mvc.OkObjectResult in 2.1447ms.
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor[1]
[2020-10-19T22:31:34.201Z]       Executing ObjectResult, writing value of type 'System.String'.
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.201Z]       Executed action Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) in 2.4314ms
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
[2020-10-19T22:31:34.201Z]       Request finished in 2.7428ms 200 text/plain; charset=utf-8
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
[2020-10-19T22:31:34.201Z]       Request starting HTTP/1.1 POST http://localhost:23890/intake/v2/events application/x-ndjson; charset=utf-8 2306
[2020-10-19T22:31:34.201Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
[2020-10-19T22:31:34.202Z]       Route matched with {action = "Post", controller = "IntakeV2Events"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] Post() on controller Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController (Elastic.Apm.Tests.MockApmServer).
[2020-10-19T22:31:34.202Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
[2020-10-19T22:31:34.202Z]       Executing action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) - Validation state: Valid
[2020-10-19T22:31:34.202Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.202Z]       Executed action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer), returned result Microsoft.AspNetCore.Mvc.OkObjectResult in 2.4247ms.
[2020-10-19T22:31:34.202Z] info: Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor[1]
[2020-10-19T22:31:34.202Z]       Executing ObjectResult, writing value of type 'System.String'.
[2020-10-19T22:31:34.202Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.202Z]       Executed action Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) in 2.6661ms
[2020-10-19T22:31:34.202Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
[2020-10-19T22:31:34.202Z]       Request finished in 2.9311ms 200 text/plain; charset=utf-8
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
[2020-10-19T22:31:34.465Z]       Request starting HTTP/1.1 POST http://localhost:23890/intake/v2/events application/x-ndjson; charset=utf-8 1475
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
[2020-10-19T22:31:34.465Z]       Route matched with {action = "Post", controller = "IntakeV2Events"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] Post() on controller Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController (Elastic.Apm.Tests.MockApmServer).
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
[2020-10-19T22:31:34.465Z]       Executing action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) - Validation state: Valid
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.465Z]       Executed action method Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer), returned result Microsoft.AspNetCore.Mvc.OkObjectResult in 2.5515ms.
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor[1]
[2020-10-19T22:31:34.465Z]       Executing ObjectResult, writing value of type 'System.String'.
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:34.465Z]       Executed action Elastic.Apm.Tests.MockApmServer.Controllers.IntakeV2EventsController.Post (Elastic.Apm.Tests.MockApmServer) in 2.8461ms
[2020-10-19T22:31:34.465Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
[2020-10-19T22:31:34.465Z]       Request finished in 3.1554ms 200 text/plain; charset=utf-8
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
[2020-10-19T22:31:35.039Z]       Request starting HTTP/1.1 GET http://localhost:23890/config/v1/agents?service.name=AspNetFullFramework_Tests_CentralConfigTests-CustomServiceName&service.environment=AspNetFullFramework_Tests_CentralConfigTests-CustomEnvironment  
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
[2020-10-19T22:31:35.039Z]       Route matched with {action = "Get", controller = "AgentsConfig"}. Executing controller action with signature Microsoft.AspNetCore.Mvc.IActionResult Get() on controller Elastic.Apm.Tests.MockApmServer.Controllers.AgentsConfigController (Elastic.Apm.Tests.MockApmServer).
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
[2020-10-19T22:31:35.039Z]       Executing action method Elastic.Apm.Tests.MockApmServer.Controllers.AgentsConfigController.Get (Elastic.Apm.Tests.MockApmServer) - Validation state: Valid
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:35.039Z]       Executed action method Elastic.Apm.Tests.MockApmServer.Controllers.AgentsConfigController.Get (Elastic.Apm.Tests.MockApmServer), returned result Microsoft.AspNetCore.Mvc.StatusCodeResult in 0.2666ms.
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Mvc.StatusCodeResult[1]
[2020-10-19T22:31:35.039Z]       Executing HttpStatusCodeResult, setting HTTP status code 304
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
[2020-10-19T22:31:35.039Z]       Executed action Elastic.Apm.Tests.MockApmServer.Controllers.AgentsConfigController.Get (Elastic.Apm.Tests.MockApmServer) in 0.4386ms
[2020-10-19T22:31:35.039Z] info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
[2020-10-19T22:31:35.039Z]       Request finished in 0.7461ms 304 
[2020-10-19T22:31:36.958Z] Elastic APM .NET Tests> 82> Console> [2020-10-19 22:31:36.616 +00:00][Info] - {LoggingTestBase} Finished test: Elastic.Apm.AspNetFullFramework.Tests.CentralConfigTests+MaxSpansAndSampleRateTests.MaxSpans_invalid_value
[2020-10-19T22:31:36.958Z] [xUnit.net 00:13:05.9221481]   Finished:    Elastic.Apm.AspNetFullFramework.Tests
[2020-10-19T22:31:36.958Z]   û Elastic.Apm.AspNetFullFramework.Tests.CentralConfigTests+MaxSpansAndSampleRateTests.MaxSpans_invalid_value [10s 604ms]
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z] JunitXML Logger - Results File: C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\junit-NETCoreApp21-Elastic.Apm.AspNetFullFramework.Tests.xml
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z] Test Run Failed.
[2020-10-19T22:31:37.220Z] Total tests: 97
[2020-10-19T22:31:37.220Z]      Passed: 96
[2020-10-19T22:31:37.220Z]      Failed: 1
[2020-10-19T22:31:37.220Z]  Total time: 13.1882 Minutes
[2020-10-19T22:31:37.220Z]      1>Project "C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\Elastic.Apm.AspNetFullFramework.Tests.csproj" on node 1 (VSTest target(s)).
[2020-10-19T22:31:37.220Z]      1>C:\Program Files\dotnet\sdk\3.1.403\Microsoft.TestPlatform.targets(32,5): error MSB4181: The "Microsoft.TestPlatform.Build.Tasks.VSTestTask" task returned false but did not log an error. [C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\Elastic.Apm.AspNetFullFramework.Tests.csproj]
[2020-10-19T22:31:37.220Z]      1>Done Building Project "C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\Elastic.Apm.AspNetFullFramework.Tests.csproj" (VSTest target(s)) -- FAILED.
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z] Build FAILED.
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z]        "C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\Elastic.Apm.AspNetFullFramework.Tests.csproj" (VSTest target) (1) ->
[2020-10-19T22:31:37.220Z]        (VSTest target) -> 
[2020-10-19T22:31:37.220Z]          C:\Program Files\dotnet\sdk\3.1.403\Microsoft.TestPlatform.targets(32,5): error MSB4181: The "Microsoft.TestPlatform.Build.Tasks.VSTestTask" task returned false but did not log an error. [C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet\test\Elastic.Apm.AspNetFullFramework.Tests\Elastic.Apm.AspNetFullFramework.Tests.csproj]
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z]     0 Warning(s)
[2020-10-19T22:31:37.220Z]     1 Error(s)
[2020-10-19T22:31:37.220Z] 
[2020-10-19T22:31:37.220Z] Time Elapsed 00:13:12.25
[2020-10-19T22:31:38.346Z] Post stage
[2020-10-19T22:31:38.366Z] Running in C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-982\apm-agent-dotnet
[2020-10-19T22:31:38.421Z] Archiving artifacts
[2020-10-19T22:31:38.841Z] Recording test results
[2020-10-19T22:31:40.384Z] Post stage
[2020-10-19T22:31:40.415Z] [WS-CLEANUP] Deleting project workspace...
[2020-10-19T22:31:40.415Z] [WS-CLEANUP] Deferred wipeout is disabled by the job configuration...
[2020-10-19T22:31:43.794Z] [WS-CLEANUP] done
[2020-10-19T22:31:44.036Z] Failed in branch Windows .NET Framework
[2020-10-19T22:31:44.534Z] Stage "Release to feedz.io" skipped due to earlier failure(s)
[2020-10-19T22:31:44.817Z] Stage "Release" skipped due to earlier failure(s)
[2020-10-19T22:31:44.878Z] Stage "Release" skipped due to earlier failure(s)
[2020-10-19T22:31:45.007Z] Stage "Release" skipped due to earlier failure(s)
[2020-10-19T22:31:45.213Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-10-19T22:31:45.266Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-10-19T22:31:45.701Z] Running on Jenkins in /var/lib/jenkins/workspace/tnet_apm-agent-dotnet-mbp_PR-982
[2020-10-19T22:31:45.844Z] [INFO] getVaultSecret: Getting secrets
[2020-10-19T22:31:46.116Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-10-19T22:31:47.253Z] + chmod 755 generate-build-data.sh
[2020-10-19T22:31:47.253Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-982/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-982/runs/1 FAILURE 3597251
[2020-10-19T22:31:47.804Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-982/runs/1/steps/?limit=10000 -o steps-info.json
[2020-10-19T22:31:48.355Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-982/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

russcam commented on src/Elastic.Apm/Model/Span.cs:
```diff
@@ -406,5 +408,23 @@ private Destination DeduceHttpDestination()
 		}
+
+		public void SetLabel(string key, string value)
+			=> Context.Labels.Add(key, value);
```
should this be assignative i.e. use the indexer?

gregkalapos replied:
Right, fixed this - and also added a test which covers this.

gregkalapos commented on src/Elastic.Apm/Model/LabelsDictionary.cs:
```diff
@@ -0,0 +1,35 @@
+	/// The reason for this is backwards compatibility - this type makes sure that we don't break user that rely on the old
+	/// interface.
+	/// </summary>
+	internal class LabelsDictionary : Dictionary<string, string>
```
Originally I thought we can shadow the `int this[string key]` indexing operator with the `new` keyword. Unfortunately this does not help, because the API surface works with `Dictionary<string,string>`.

On `Transaction.cs` we have:

```
public Dictionary<string, string> Labels => Context.Labels;
````

and `Context.Labels` is the instance of this `LabelsDictionary`. 

But if we use the `new` keyword to shadow, then `Transaction.Labels["foo"] = "bar"` will hit the `int this[T key]` operator of `Dictionary` and not the one that we'd implement here. 

So, the only option I could come up with is to have 2 dictionaries: 1 for the items used through the `Dictionary<string,string>` API and 1 for the ones added through the `SetLabel` method - and we merge them before serialization.

gregkalapos commented:
@russcam this is ready for another review.

The `IIS Tests` stage in CI does not seem to run - it's queued for a Windows machine since ~15 minutes and nothing happens. I suspect this is a CI-infra issue. All other stages are green.

bmorelli25 commented on docs/public-api.asciidoc:
```diff
@@ -248,9 +248,45 @@ finally
+
+Label values can be a string, boolean, or number.
+Because labels for a given key are stored in the same place in Elasticsearch, all label values of a given key must have the same data type.
+Multiple data types per key will throw an exception, e.g. `{foo: bar}` and `{foo: 42}`
```
It looks like you copied this from the Java Agent, who copied it from the APM Overview. . . and they forgot the period 😅 
```suggestion
Multiple data types per key will throw an exception, e.g. `{foo: bar}` and `{foo: 42}`.
```

bmorelli25 commented on docs/public-api.asciidoc:
```diff
@@ -248,9 +248,45 @@ finally
 
+[float]
+[[api-transaction-set-label]]
+==== `void SetLabel(string key, T value) added[1.7.0,Number and boolean labels require APM Server 6.7]
```
Maybe add a plus here (and also the closing backtick)

```suggestion
==== `void SetLabel(string key, T value)` added[1.7.0,Number and boolean labels require APM Server 6.7+]
```

bmorelli25 commented on docs/public-api.asciidoc:
```diff
@@ -473,10 +509,37 @@ finally
 [[api-span-tags]]
 ==== `Dictionary<string,string> Labels`
+
+WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-span-set-label>> method instead.   
```
This looks funky when rendered. Let's add a short title to make it cleaner
```suggestion
WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-span-set-label,`void SetLabel()`>> method instead.   
```

bmorelli25 reviewed (approved):
A few recommendations, then docs LGTM.

russcam commented on docs/public-api.asciidoc:
```diff
@@ -248,9 +248,45 @@ finally
 [[api-transaction-tags]]
 ==== `Dictionary<string,string> Labels`
+
+WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-transaction-set-label>> method instead.   
```
```suggestion
WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-transaction-set-label, `void SetLabel()`>> method instead, which allows setting labels of string, boolean and number. This property remains for now in order to not break binary compatibility, and at serialization time, the values set with `.SetLabel()` are combined with `Labels` to form the set of labels sent to APM server, with values in `Labels` taking precedence. 
```

russcam commented on docs/public-api.asciidoc:
```diff
@@ -473,10 +509,37 @@ finally
 [[api-span-tags]]
 ==== `Dictionary<string,string> Labels`
+
+WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-span-set-label,`void SetLabel()`>> method instead.   
```
```suggestion
WARNING: This property is obsolete and will be be removed in a future version. Use the <<api-span-set-label, `void SetLabel()`>> method instead, which allows setting labels of string, boolean and number. This property remains for now in order to not break binary compatibility, and at serialization time, the values set with `.SetLabel()` are combined with `Labels` to form the set of labels sent to APM server, with values in `Labels` taking precedence. 
```

russcam commented on test/Elastic.Apm.AspNetCore.Tests/AspNetCoreDiagnosticListenerTest.cs:
```diff
@@ -53,20 +52,20 @@ public async Task TestErrorInAspNetCore(bool useOnlyDiagnosticSource)
 				capturedPayload.Should().NotBeNull();
 
-				capturedPayload.Transactions.Should().ContainSingle();
+				capturedPayload?.Transactions.Should().ContainSingle();
```
The null conditionals aren't needed here IMO, as a null reference will fail the test on [line 53](https://github.com/elastic/apm-agent-dotnet/pull/982/files#diff-7c522236e53c6b6553577390af5e0b16cc269c08530571ec49216b7017ebecd9R53)

gregkalapos replied:
I totally agree with you, but Rider is not smart enough and gives a warning if we don't care about the potential `null`:

![image](https://user-images.githubusercontent.com/1091853/97860642-ca884380-1d02-11eb-8bd5-7ac3c0787105.png)

I'd keep the `?.` that way we don't have warnings.

russcam replied:
unfortunate to code in a specific way to satisfy tooling 😃 I had a quick look at seeing if null-checking options can be configured for this, but doesn't look so. Maybe wrapping all in a `if(capturedPayload != null) { ... }` satisfies the warning _and_ only does the check once?

gregkalapos replied:
Using `if` block would definitely work - but that's also coding in a specific way to satisfy tooling - Ultimately I would not null check here (as we agree it's not needed), but the warnings drive me crazy :)

I don't have strong feelings and I'd be happy to push a change to use `if`s - only thing is that I use `?.` for this in lots of other places.

Again, I'm fine with both `?.` and `if` blocks, but mixing the two would be maybe nice to avoid. Would you be ok with maybe changing this in a separate PR at all places in case we really want to do so?

russcam replied:
for sure! Totally not needed for this PR

russcam commented on test/Elastic.Apm.AspNetCore.Tests/AspNetCoreDiagnosticListenerTest.cs:
```diff
@@ -90,24 +89,24 @@ public async Task TestJsonBodyRetrievalOnRequestFailureInAspNetCore(bool useOnly
 				capturedPayload.Should().NotBeNull();
 
-				capturedPayload.Transactions.Should().ContainSingle();
+				capturedPayload?.Transactions.Should().ContainSingle();
```
The null conditionals aren't needed here IMO, as a null reference will fail the test on [line 94](https://github.com/elastic/apm-agent-dotnet/pull/982/files#diff-7c522236e53c6b6553577390af5e0b16cc269c08530571ec49216b7017ebecd9R94)

gregkalapos replied:
Same as in https://github.com/elastic/apm-agent-dotnet/pull/982#discussion_r515686044

russcam commented on src/Elastic.Apm/Api/Context.cs:
```diff
@@ -4,25 +4,29 @@
 
 namespace Elastic.Apm.Api
 {
 	public class Context
```
Should `Context` expose `SetLabel()` methods in order to set labels of different types, that `Transaction.SetLabel()` methods should call? In other words, are there usages where labels might be set on `Context` outside of the scope of a `Transaction`? There are some in unit tests

gregkalapos replied:
> Should Context expose SetLabel() methods in order to set labels of different types, that Transaction.SetLabel() methods should call? In other words, are there usages where labels might be set on Context outside of the scope of a Transaction?

I don't think so - there are 2 use cases when you set labels for `Context`:

- with transactions and for those we have public `SetLabels()` methods
- for errors and in that case we also have a public API to pass labels: `ITransaction.CaptureError("Test Error", "TestMethod", new StackTrace().GetFrames(), labels: new Dictionary<string, Label> { { "stringLabel", "test" } });` . In case of errors the `Context` isn't exposed, because everything is filled when the error is captured on the fly - so no way to change an `Error.Context` after an error happened (expect with a filter).

I think exposing `Labels` on the `Context` wasn't a good idea because it just duplicates the API that we already have on `ITransaction` (and on the `CaptureError`/`CaptureException` methods) which is confusing: it's literally the same value with a duplicated API.

So overall, I think the cleanest would be to just keep the `.Labels` property with the obsolete attribute for backwards compatibility and then remove the whole Label thing from `Context`. 

> There are some in unit tests

I think I rewrote all of them to use the new public API (except the ones for backwards compatibility). In `SerializationTests` we have tests that new up a `Context` and serialize it and in those we use the new `internal InternalLabels` property. Those unit tests aren't real API use cases, in those cases there is no `Transaction` or `Error`, the tests only care about the serialization.

russcam replied:
got it, thanks

russcam commented on src/Elastic.Apm/Api/Context.cs:
```diff
@@ -4,25 +4,29 @@
 		/// <seealso cref="ShouldSerializeLabels" />
 		/// </summary>
 		[JsonProperty("tags")]
 		[JsonConverter(typeof(LabelsJsonConverter))]
```
I don't think this should have converter type `LabelsJsonConverter`. The `[JsonConverter(typeof(LabelsJsonConverter))]` can be applied to `LabelsDictionary` and removed from label properties on `Context`, `Metadata`, `SpanContext` e.g.

```c#
	[JsonConverter(typeof(LabelsJsonConverter))]
	internal class LabelsDictionary : Dictionary<string, string>
    {
        // ...
    }
```

It achieves the same outcome and removes the need to attribute every labels property

gregkalapos replied:
Makes sense - I implemented it. Only thing is that if we do this then we'll serialize labels through the `InternalLabels` property instead of the `Labels` property and `InternalLabels` is `Lazy<LabelsDictionary>`  which again needs another converter. But overall I think it's still better that way.

russcam replied:
I think we might be talking about slightly different things. I suggested to move the `[JsonConverter(typeof(LabelsJsonConverter))]` on all label properties to the `LabelsDictionary` type. With this change, wherever `LabelsDictionary` is used, the converter will be used. I _think_ you're talking about attributing `InternalLabels`, which I don't think we should do, or need to do. The public API of the `"tags"` property is still `Dictionary<string, string>`, but because the specific implementation of the instance returned is `LabelsDictionary` that has an associated `JsonConverter`, the converter will be used. I think it's a little unituitive to be honest, but allows keeping the lazy semantics and moves the association of `LabelsDictionary` and `LabelsJsonConverter` to one place.

gregkalapos replied:
Oh, I think now I got it. I just pushed a commit: https://github.com/elastic/apm-agent-dotnet/pull/982/commits/c1eb1680418b85f7b193cce051aebaa6e3082917 

Is that what you mean? 

I kind of agree it's a bit unituitive and the other thing is that we'll serialize labels through an obsolete property. But overall I'm fine with it. Plus I agree it's nice to have the association of `LabelsDictionary` and `LabelsJsonConverter` at one place

russcam commented on src/Elastic.Apm/Api/Label.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Elastic.Apm.Api
+{
+	public class Label
```
Should add some XML documentation to indicate what this represents, and highlight the implicit converion from supported values

gregkalapos replied:
Agreed - added XML doc.

russcam commented on src/Elastic.Apm/Model/Span.cs:
```diff
@@ -406,5 +412,23 @@ private Destination DeduceHttpDestination()
 			}
 		}
+
+		public void SetLabel(string key, string value)
```
I wonder if we should have only a single `SetLabel(string key, Label value)` method, and delegate the "overloads" to the supported implicit conversions of `Label`. It would alleviate the need to maintain 6 overloads in all places that support setting labels. The negative is the somewhat opaqueness of implicit conversions, with the overloads acting to promote the implicit conversions supported. 

Both the following would work with a single `SetLabel(string key, Label value)` method

```c#
SetLabel("foo", "bar");
```

and

```c#
SetLabel("foo", new Label("bar"));
```

gregkalapos replied:
Hmm... I personally prefer having baked in types on the API surface. If people see `SetLabel(string, bool)` and the other overloads they'll immediately know what types they can pass in. I'd keep it for now.

russcam replied:
ok. Definitely more discoverable with overloads. I've maybe been thinking too much in Rust lately!

russcam commented on src/Elastic.Apm/Api/SpanContext.cs:
```diff
@@ -5,14 +5,15 @@
 
 namespace Elastic.Apm.Api
 {
 	public class SpanContext
```
Should `SpanContext` expose `SetLabel()` methods in order to set labels of different types, that `Span.SetLabel()` methods should call? In other words, are there usages where labels might be set on `SpanContext` outside of the scope of a `Span`? There are some in unit tests

gregkalapos replied:
Same as https://github.com/elastic/apm-agent-dotnet/pull/982#discussion_r515903177

russcam reviewed (commented):
I left some comments for consideration

russcam commented on test/Elastic.Apm.Tests/LabelTests.cs:
```diff
@@ -0,0 +1,287 @@
+
+namespace Elastic.Apm.Tests
+{
+	public class LabelTests
```
The tests should use the payload serializer instead of `JsonConvert.SerializeObject()`, It better represents the agent's serialization and means the tests don't need to change when swapping to a different serializer in the future

gregkalapos replied:
Makes sense. Turns out we have `SerializationTests.cs` which does exactly that - now `LabelTests.cs` does it as well.

russcam commented on src/Elastic.Apm/Report/Serialization/LabelsJsonConverter.cs:
```diff
@@ -23,16 +23,26 @@ public override void WriteJson(JsonWriter writer, Dictionary<string, string> lab
 			bool hasExistingValue, JsonSerializer serializer
 		)
-			=> serializer.Deserialize<Dictionary<string, string>>(reader);
+			=> serializer.Deserialize<LabelsDictionary>(reader);
```
This causes a stackoverflow as `Deserialize<LabelsDictionary>` ends up calling the converter's `ReadJson` implementation

russcam commented on src/Elastic.Apm/Report/Serialization/LabelsJsonConverter.cs:
```diff
@@ -23,16 +23,26 @@ public override void WriteJson(JsonWriter writer, Dictionary<string, string> lab
 			bool hasExistingValue, JsonSerializer serializer
 		)
-			=> serializer.Deserialize<Dictionary<string, string>>(reader);
+			=> serializer.Deserialize<LabelsDictionary>(reader);
```
```suggestion
		{
			if (reader.TokenType == JsonToken.Null)
				return null;

			if (reader.TokenType != JsonToken.StartObject)
				throw new JsonException($"expected {JsonToken.StartObject} but received {reader.TokenType}");

			var labels = new LabelsDictionary();

			while (reader.Read())
			{
				if (reader.TokenType == JsonToken.EndObject)
					break;

				var property = (string)reader.Value;
				reader.Read();

				switch (reader.TokenType)
				{
					case JsonToken.Integer:
						labels.InnerDictionary.Add(property, (long)reader.Value);
						break;
					case JsonToken.Float:
						labels.InnerDictionary.Add(property, (double)reader.Value);
						break;
					case JsonToken.String:
						labels.Add(property, (string)reader.Value);
						break;
					case JsonToken.Boolean:
						labels.InnerDictionary.Add(property, (bool)reader.Value);
						break;
					default:
						throw new JsonException(
							$"Expected {JsonToken.Integer}, {JsonToken.Float}, {JsonToken.String} or {JsonToken.Boolean} "
							+ $"but received {reader.TokenType}");
				}
			}

			return labels;
		}
```

russcam replied:
One question that came to mind - do labels need to support dates? Understand that they will be serialized as a JSON string, but wondering if deserialization has to consider `JsonToken.Date`

gregkalapos replied:
> do labels need to support dates? 

No, my understanding is that no other agent supports dates. It's only string, numbers, and bool. I guess technically we could support dates, but so far I followed other agents and those don't support it. Before we add it I'd definitely see if there is feedback and need for it at al.

>  wondering if deserialization has to consider `JsonToken.Date`

Since the de-serialization logic is not used in production and I feel date isn't needed to be supported I think it's fine to ignore it. As discussed over zoom, the reason I'm not very worried about the de-serialization logic is that the agent only uses the serialization logic to go from C# objects to JSON, but not the other way, so if de-serialization is important at al, then it's maybe in some of our internal tests scenarios (but currently that's not even a case based on what we discussed (Mock APM Server e.g. does not use this)).

russcam reviewed (changes requested):
I had another review and left some comments.

I had a brief look and don't think we have tests for deserializing `LabelsDictionary` either directly, or via `Context` (which would test the deserialization through the `Dictionary<string,string>` property type). It might be prudent to have these tests

gregkalapos commented:
> 
> 
> I had another review and left some comments.
> 
> I had a brief look and don't think we have tests for deserializing `LabelsDictionary` either directly, or via `Context` (which would test the deserialization through the `Dictionary<string,string>` property type). It might be prudent to have these tests

I did another round on this PR. 

- As suggested now we test with `PayloadItemSerializer` instead of `JsonConvert.SerializeObject`. 
- Kept the `LabelsJsonConverter.ReadJson` method - I took the suggested implementation. As discussed there is no production use-case for it - we never serialize in a real prod. agent. 
- Also added a full framework test which tests the labels and I also switched the `.Labels` to use `LabelsDictionary` there. With this at least in the MockApmServer we use the de-serialization logic, plus the API is also closer to the real agent with that change.

russcam commented on test/Elastic.Apm.Tests/LabelTests.cs:
```diff
@@ -0,0 +1,295 @@
+				{
+					var labelName = "myLabel";
+					SetLabel(s, labelValue, labelName);
+					var jsonString = new PayloadItemSerializer(new MockConfigSnapshot()).SerializeObject(s);
```
```suggestion
					var jsonString = SerializePayloadItem(s);
```

gregkalapos replied:
Oh, good catch - this is a leftover from our zoom discussion and replace-all of course did not find this.

russcam reviewed (approved):
LGTM - great effort, @gregkalapos 👍 

I left one small suggestion to use the serialize method in one of the tests
